src/test/java/org/fogbowcloud/manager/core/processors/FulfilledProcessorTest.java

```diff
@@ -99,355 +95,365 @@ public void tearDown() {
-
-        Assert.assertEquals(this.connectionAttempts.size(), 0);
-    }
-    
+    // test case: When running thread in FulfilledProcessor without the external addresses of the
+    // reverse tunnel, the method processFulfilledOrder() must not change OrderState to Failed and
+    // must remain in Fulfilled list.
+    @SuppressWarnings("static-access")
     @Test
     public void testRunProcessLocalComputeOrderWithoutExternalAddresses()
             throws FogbowManagerException, UnexpectedException, InterruptedException {
-        
+
+        // set up
         Order order = this.createOrder();
         order.setOrderState(OrderState.FULFILLED);
         this.fulfilledOrderList.addItem(order);
+        Assert.assertNull(this.failedOrderList.getNext());
 
         Instance orderInstance = Mockito.spy(new ComputeInstance(FAKE_INSTANCE_ID));
         orderInstance.setState(InstanceState.READY);
         order.setInstanceId(FAKE_INSTANCE_ID);
 
         mockCloudConnectorFactory(orderInstance);
 
+        // exercise
         Mockito.when(this.tunnelingService.getExternalServiceAddresses(Mockito.anyString()))
                 .thenReturn(null);
 
-        Assert.assertNull(this.failedOrderList.getNext());
-
         this.thread = new Thread(this.fulfilledProcessor);
         this.thread.start();
+        this.thread.sleep(DEFAULT_SLEEP_TIME);
 
-        Thread.sleep(MIN_SLEEP_TIME);
+        // verify
+        Mockito.verify(this.tunnelingService).getExternalServiceAddresses(Mockito.anyString());
 
-        Order test = this.fulfilledOrderList.getNext();
-        Assert.assertNotNull(test);
-        Assert.assertEquals(order.getInstanceId(), test.getInstanceId());
-        Assert.assertEquals(OrderState.FULFILLED, test.getOrderState());
+        Assert.assertNotNull(this.fulfilledOrderList.getNext());
+        Assert.assertNull(this.failedOrderList.getNext());
     }
 
+    // test case: When running thread in the FulfilledProcessor and the OrderType is not a
+    // Compute, the method processFulfilledOrder() must not change OrderState to Failed and
+    // must remain in Fulfilled list.
     @Test
     public void testRunProcessLocalWhenOrderTypeIsNotComputer()
             throws FogbowManagerException, UnexpectedException, InterruptedException {
 
+        // set up
         FederationUser federationUser = Mockito.mock(FederationUser.class);
         String providingMember = LOCAL_MEMBER_ID;
-        
         String requestingMember =
                 String.valueOf(this.properties.get(ConfigurationConstants.XMPP_JID_KEY));
 
         Order order = new AttachmentOrder(federationUser, requestingMember, providingMember,
                 FAKE_SOURCE, FAKE_TARGET, FAKE_DEVICE);
-        
+
         order.setOrderState(OrderState.FULFILLED);
-        
+
         this.fulfilledOrderList.addItem(order);
+        Assert.assertNull(this.failedOrderList.getNext());
 
         Instance orderInstance = Mockito.spy(new ComputeInstance(FAKE_INSTANCE_ID));
```

Comment:
Do we really need to spy this object?

Reply:
No, I will remove the spy from this object.